A method to plot beam diagram using sympy's own plot()

ishanaj opened the pull request:
This PR is an initiative to find an alternative to PR #17240 . With both PR's having the same goal. This PR intends to directly use `sympy.plot()` rather than using matplotlib or the `backend` attribute.
Currently the `draw()` function in this PR can plot:
- [x] rectangle for the beam 
- [x] loads ~~except moment and point loads~~
- [x] supports
- [x] arrows for the loads
- [x] arrows (or colour fill) for uniformly varying loads and other higher order loads
- [x] scaling of the plot
- [x] documentation


#### Release Notes

<!-- BEGIN RELEASE NOTES -->
- physics.continuum_mechanics
    - a new method `draw()` has been added in the beam module which gives functionality to draw the beam diagrams using matplotlib
-  plotting
    -  functionalities of adding markers, annotations, rectangles, filling color in a plot have been added to the matplotlib backend of the plotting module
<!-- END RELEASE NOTES -->

ishanaj committed:
a method to plot beam diagram using sympy's own plot()

ishanaj commented:
Tested the following example:
```
>>> E, I = symbols('E, I')
>>> b = Beam(9, E, I)
>>> b.apply_load(-12, 9, -1)  # gets skipped
>>> b.apply_load(50, 5, -2)  # gets skipped
>>> b.apply_load(3, 6, 1, end=8)
>>> b.apply_load(4, 0, 0, end=5)
>>> b.draw()
```
![Screenshot - 05-08-2019 , 19_49_21](https://user-images.githubusercontent.com/29159034/62473406-4c1aba00-b7be-11e9-84ec-76a6af48a10a.png)

I have a few questions to discuss:
- How to put arrows for these loads or fill it with color? I have tried `fill=True` but that doesn't work as probably there is no closed area.
- Also, currently the magnitude of the load is low, but if it gets high we might need to scale down the height of the plot of uniform load and varying load. I have tried `yscale` but not able to get how that works. Any ideas?

ping @jashan498 @moorepants

moorepants commented:
> How to put arrows for these loads or fill it with color? I have tried fill=True but that doesn't work as probably there is no closed area.

You'll have to add that to the sympy.plot() features if it doesn't already exist. Or you add them manually with matplotlib in the beam code. The ideal thing is to enhance sympy.plot() so that it benefits all sympy plotting needs. For example, adding a common way to plot y = 1 @ x = 0, y = 0 x!=0 would be a nice addition.

> Also, currently the magnitude of the load is low, but if it gets high we might need to scale down the height of the plot of uniform load and varying load. I have tried yscale but not able to get how that works. Any ideas?

Good point. There is the idea of plotting an accurate load curve and the idea of making a pleasing pictorial version of the beam. For the latter, you may need to take liberties to scale so the beam always looks nice (like in text books). Maybe both should ultimately be supported. For example you could do `Beam.draw()` for the realistic one and `Beam.draw(pictorial=True)` for the scaled version.

ishanaj committed:
added features in the plotting module for plotting annotations, markers, and
rectangles, and modified the beam.draw() function to use these to plot
the beam diagram

ishanaj commented:
I have added some features in the plotting module using which we can now use annotations, markers and rectangle via `sympy.plot`. I will be adding documentation and test for both the modules later.

Also, annotations, markers, rectangles are used as keyword arguments in the plot function. They are in fact a list of the dictionaries, **eg**:
**`annotations=[{'a': x, 'b': y........}, {'a2': x2, 'b2': x2.....},....]`**
 where each dictionary represents all the parameters that are to be passed in the `ax.annotate()`  function in the backend

I have tested the following example:
```
R1, R2 = symbols('R1, R2')
E, I = symbols('E, I')

b1 = Beam(50, 20, 30)

b1.apply_load(10, 2, -1)
b1.apply_load(R1, 10, -1)
b1.apply_load(R2, 30, -1)
b1.apply_load(1, 5, 0, 23)
b1.apply_load(1, 30, 1, 50)
b1.apply_support(50, "pin")
b1.apply_support(0, "fixed")
b1.apply_support(20, "roller")
b1.draw()
```

![Screenshot - 08-08-2019 , 21_48_25](https://user-images.githubusercontent.com/29159034/62723091-7a520100-ba2d-11e9-9830-c3ad313ea342.png)

moorepants commented:
Looks nice!

ishanaj committed:
added functionality of filling colour in plot using matplotlib's
fill_between(), and used the same in representing higher order beam loads

ishanaj commented:
The diagram now looks like:
![image](https://user-images.githubusercontent.com/29159034/62806483-9d091600-bb10-11e9-8a1d-1ce4253d2fa5.png)

I have considered coloring the higher loads (rather than filling them with arrows) as it would further make the plot clean in cases where a point load is superimposed on them.

ishanaj committed:
added functionality to scale the beam diagram using a parameter
pictorial=True, also modified plot.markers a bit

ishanaj commented:
I have added the option for the user to scale the plot via a parameter `pictorial=True`.
This would be helpful in the cases where the magnitude of the load is too high (or too low) and might look a bit indecent or may go out of the plot.
This is an example that has a similar problem:
```
>>> R1, R2 = symbols('R1, R2')
>>> E, I = symbols('E, I')

>>> b1 = Beam(50, 20, 30)

>>> b1.apply_load(10, 2, -1)
>>> b1.apply_load(R1, 10, -1)
>>> b1.apply_load(R2, 30, -1)
>>> b1.apply_load(90, 5, 0, 23)
>>> b1.apply_load(10, 30, 1, 50)
>>> b1.apply_support(50, "pin")
>>> b1.apply_support(0, "fixed")
>>> b1.apply_support(20, "roller")
# case 1 on the left
>>> p = b1.draw()
>>> p.show()

# case 2 on the right
>>> p1 = b1.draw(pictorial=True)
>>> p1.show()
```
![Screenshot - 10-08-2019 , 23_04_45](https://user-images.githubusercontent.com/29159034/62825282-fa5f9e80-bbc6-11e9-88bc-a67cca1c9abf.png)

ishanaj committed:
Merge branch 'master' into beam_diagram_2

ishanaj committed:
added documentation for draw function and the plot features

moorepants commented:
If you are plotting the unit of length on the x axis and unit of force on the y axis, what is your default scaling factor, 1:1? Does it make sense to do 1:1 as a default? I'm not sure it does. What is the exact difference in the pictorial or not?

ishanaj commented:
Setting `pictorial=True` would simply alter the height of the udl, vdl or higher loads and the rest would remain same. Here we set udl magnitude to length/2, and vdl and other higher order loads magnitude/unit to 1N/m, for any plot.

`pictorial =False` makes the diagram with the exact dimensions, so a scale of 1:1 is present.
But in some cases the magnitude of the load becomes high due to which the height of the udl or vdl gets high enough to go out of the plot as the ylimit is set to `-length, length`. So as to decrease this height to a nominal amount such that the plot looks better we can use `pictorial=True`

ishanaj committed:
corrected failing tests

ishanaj committed:
removed extra imports and idented example in the docs which was causing
the build fail

ishanaj commented:
The build fails for python 3.7. I don't know why it is not able to import numpy using`import_kwargs`.
Initially, I was running it on python 3.5 and everything works well with it. 
Any idea why it does so?

asmeurer commented:
Your tests need to correctly skip when numpy is not installed. NumPy is an optional dependency, so the tests should not fail when it is not installed.

ishanaj committed:
skipped doctest and raised ImportError if numpy is not imported

ishanaj commented:
@asmeurer I have added a line in the function which checks whether numpy is successfully imported and raises import error accordingly.
Also I have skipped the doctests currently, but is there a way to skip them only when numpy is not installed?

jashan498 commented:
You can use `@doctest_depends_on(modules=('numpy'))` decorator to skip doctests in such scenarios.

ishanaj committed:
used doctest_depends_upon to skip test when numpy not installed

ishanaj commented:
ping @moorepants @jashan498 The work on this PR is complete. Please have a look

ishanaj committed:
corrected failing test by making plot handle case of empty plot case

ishanaj commented:
The tests have finally passed!

oscargus commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -1521,6 +1524,137 @@ def plot_loading_results(self, subs=None):
+        markers += support_markers
+
+        sing_plot = plot(height + load_eq, (x, 0, length),
+         xlim=(-2, length + 2), ylim=(-length, length), annotations=annotations,
```
Would it make sense to use relative spacing instead of absolute? Like `(-0.05*length, 1.05*length)`?

ishanaj replied:
I tried `xlim=(-0.05*length, 1.05*length)`, it more or less gives similar output.

The main aim of having `xlim` is to include the rectangle (denoting the fixed support) in the plot.
Since the width of this rectangle is every time `length/10` (see line 1651), I think then we can rather go for `xlim=(-length/10, length + length/10)`.
I guess then even this would now be relative. 
Should we go for this then? @oscargus

ishanaj committed:
Merge branch 'master' into beam_diagram_2

ishanaj commented:
I guess after this we should also handle the case where the length is a symbol. Just adding a single line like :`length = 10 if isinstance(self.length, Symbol) else self.length` i.e setting a default length, would work.
Also, we can handle the case where the magnitude of load is a variable.
This would be automatically handled when `pictorial=-True`, but for the case where `pictorial=False`
either we can reset it to true or we could urge the user to try again using `pictorial=True`.
Also, if there is no need for 1:1 scaling (i.e pictorial=False), we can remove it. If we do so then we don't have to worry about the case where magnitude is a symbol.
@jashan498 @moorepants Should we be removing 1:1 scaling (i.e., pictorial=False)?

jashan498 commented:
Not sure about removing `pictorial=False` but you can set `pictorial=True` as default instead and return an error if someone uses a `Symbol` with `pictorial=False`.

ishanaj committed:
made plot limits relative, handled the case where length and magnitude
of loads are symbols

jashan498 commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -1519,6 +1524,137 @@ def plot_loading_results(self, subs=None):
+                    load_args = scaled_load.args
+                else:
+                    if isinstance(load[0], Symbol):
+                        raise ValueError("Magnitude of load: {} should not be a Symbol".format(load))
```
This error message seems to be a little misleading. It can be a `Symbol` when `Pictorial` is set to `True`.

ishanaj replied:
How about this:
`"Magnitude of load: {} should not be a Symbol when pictorial is False. Try setting pictorial to True".format(load)`

jashan498 replied:
Looks good!

jashan498 commented:
ping @ishanaj

ishanaj committed:
corrected the error message

ishanaj commented:
@jashan498 The build has errored. Could you restart the build?

jashan498 commented:
Restarted!

jashan498 commented:
Restarted

ishanaj commented:
The work on this PR is done and I think it is now ready to go. @moorepants @jashan498

jashan498 commented:
I'm getting this error while plotting a beam with no load applied on it.
```
>>> b = Beam(50, 20, 30)
>>> b.apply_support(0, "fixed")
>>> b.draw()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jashan/sympy/sympy/physics/continuum_mechanics/beam.py", line 1577, in draw
    annotations, markers, load_eq, fill = self._draw_load(pictorial, length)
  File "/home/jashan/sympy/sympy/physics/continuum_mechanics/beam.py", line 1629, in _draw_load
    expr = height + load_eq.rewrite(Piecewise)
UnboundLocalError: local variable 'load_eq' referenced before assignment
```

moorepants commented:
Does this work with composite beams?

ishanaj commented:
> Does this work with composite beams?

I haven't tried them, but I think it should. I will be testing some examples and will let you know.

ishanaj committed:
made draw() accept length as a symbol for both simple and composite beams. Also optimized parts of the code.

ishanaj commented:
I have made some changes in the code. Previously, if the length was a Symbol, the value of length was not being substituted in the **position of the load** and hence the above error came. Now it has been handled.
Also, the case where the length of the composite beam is an expression in terms of Symbol has been handled. (e.g. `b.length = l+k` where b is a composite beam made by joining 2 beams of length `l` and `k`)

>  x = symbols('x')
b2 = Beam(x, 20, 30)
 b2.apply_load(90, 5, 0, 23)
b2.apply_support(0, "fixed")
b2.draw()
...........
......
.......
TypeError: can't convert expression to float

Also, the example above is a bit of problematic itself.
I think in this example, it should apply loads at positions in terms of the length `x` and not any random integer. Also, we might want to change the `base_char` to some other variable as it also by default becomes `x`.
But still, if we change it to some other variable (say `l`), then now it would not throw an error it would rather plot a wrong diagram, as in such cases where beam length is a symbol, the draw() function picks a default value of 10 for the length, but if the position of the load is an integer greater than 10 then it goes out of the plot window.

ishanaj commented:
The draw() function does work for composite beams. But there is no indication of a joint or any point where the next beam starts.

jashan498 commented:
> Also, we might want to change the base_char to some other variable as it also by default becomes x.

This issue is not just limited to plotting. We shouldn't let same symbol represent both: `variable` and `length` of the Beam object. Wrong results are returned if that happens:
```
>>> x = symbols('x')
>>> b = Beam(x, 20, 30)     # x represnts both length & variable
>>> b.apply_load(20, 0, 0, end=x/2)
>>> b.bc_deflection = [(0, 0)]
>>> b.bc_slope = [(0, 0)]
>>> b.deflection() 
          4
        x  
   4   <─> 
<x>     2  
──── - ────
720    720 

```
```
>>> l = symbols('l')
>>> b = Beam(l, 20, 30)       
>>> b.apply_load(20, 0, 0, end=l/2)
>>> b.bc_deflection = [(0, 0)]
>>> b.bc_slope = [(0, 0)]
>>> b.deflection()              # correct value
       3        4                   4
   -l       -l                l      
x⋅<───>    <───>       4   <- ─ + x> 
    2        2      <x>       2      
──────── + ────── + ──── - ──────────
  180       720     720       720    

```

ishanaj commented:
> This issue is not just limited to plotting. We shouldn't let same symbol represent both: variable and length of the Beam object. Wrong results are returned if that happens.

I think then we should be making a separate issue or PR to resolve this.

ishanaj commented:
I have opened a new issue #17602 regarding the above problem.
Also we can now move ahead with this PR.

jashan498 commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -1519,6 +1524,166 @@ def plot_loading_results(self, subs=None):
+            >>> b.apply_support(50, "pin")
+            >>> b.apply_support(0, "fixed")
+            >>> b.apply_support(20, "roller")
+            >>> b.draw(pictorial=True)
```
`pictorial` is set to `True` by default, no need to pass it in the docstring.

jashan498 commented:
Im geting an error if I don't pass `end` paramter while applying load. It should be set to end of the beam if not passed.
```
>>> b = Beam(50, 20, 30)
>>> b.apply_load(10, 2, 0)
>>> b.draw()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jashan/sympy/sympy/physics/continuum_mechanics/beam.py", line 1587, in draw
    annotations, markers, load_eq, fill = self._draw_load(pictorial, length, l)
  File "/home/jashan/sympy/sympy/physics/continuum_mechanics/beam.py", line 1643, in _draw_load
    SingularityFunction(x, end, i) / factorial(i))
TypeError: unsupported operand type(s) for -: 'NoneType' and 'Integer'

```

ishanaj committed:
removed error which occured in draw() function when end condition of the load was not applied, corrected documentation and added comments

jashan498 commented:
Diagrams look great! Merging

ishanaj commented:
Thanks for merging!
